src/databricks/labs/blueprint/paths.py

```diff
@@ -448,17 +441,296 @@ def match(self, path_pattern, *, case_sensitive=None):
+        errors: str | None = None,
+        newline: str | None = None,
+    ):
+        """Open a DBFS file.
+
+        Only text and binary I/O are supported in basic read or write mode, along with 'x' to avoid overwriting."""
+        is_write = "w" in mode
+        is_read = "r" in mode or not is_write
+        if is_read and is_write:
+            msg = f"Unsupported mode: {mode} (simultaneous read and write)"
+            raise ValueError(msg)
+        is_binary = "b" in mode
+        is_text = "t" in mode or not is_binary
+        if is_binary and is_text:
+            msg = f"Unsupported mode: {mode} (binary and text I/O)"
+            raise ValueError(msg)
+        is_overwrite = is_write and "x" not in mode
+        binary_io = self._ws.dbfs.open(self.as_posix(), read=is_read, write=is_write, overwrite=is_overwrite)
+        if is_text:
+            return io.TextIOWrapper(binary_io, encoding=encoding, errors=errors, newline=newline)
+        return binary_io
+
+    def write_bytes(self, data):
+        """Write the (binary) data to this path."""
+        # The DBFS BinaryIO implementation only accepts bytes and rejects the (other) byte-like builtins.
+        match data:
+            case builtins.bytes | builtins.bytearray:
+                binary_data = bytes(data)
+            case _:
+                binary_data = bytes(memoryview(data))
+        with self.open("wb") as f:
+            return f.write(binary_data)
+
+    @property
+    def _file_info(self) -> FileInfo:
+        # this method is cached because it is used in multiple is_* methods.
+        # DO NOT use this method in methods, where fresh result is required.
+        try:
+            return self._cached_file_info
+        except AttributeError:
+            self._cached_file_info = self._ws.dbfs.get_status(self.as_posix())
+            return self._cached_file_info
+
+    def is_dir(self) -> bool:
+        """Return True if the path points to a DBFS directory."""
+        try:
+            return bool(self._file_info.is_dir)
+        except DatabricksError:
+            return False
+
+    def is_file(self) -> bool:
+        """Return True if the path points to a file in Databricks Workspace."""
+        return not self.is_dir()
+
+    def expanduser(self: P) -> P:
+        # Expand ~ (but NOT ~user) constructs.
+        if not (self._drv or self._root) and self._path_parts and self._path_parts[0][:1] == "~":
+            if self._path_parts[0] == "~":
+                user_name = self._ws.current_user.me().user_name
+            else:
+                other_user = self._path_parts[0][1:]
+                msg = f"Cannot determine home directory for: {other_user}"
+                raise RuntimeError(msg)
+            if user_name is None:
+                raise RuntimeError("Could not determine home directory.")
+            homedir = f"/Users/{user_name}"
+            return self.with_segments(homedir, *self._path_parts[1:])
+        return self
```

Comment:
DBFS per se does not have any concepts of homedirs (or any security), so it's better to keep this method only in the parent class. we'll emulate homedirs in DBFS that way.

Reply:
Fixed; `.expanduser()` is now implemented on the base class.